### PR TITLE
allow a new upgrade version without DB changes to be inserted

### PR DESCRIPTION
**Pull request by DaanHoogland**
### Description

This PR...

Fixes: #6760

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### How Has This Been Tested?

unit tests (so far)

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@blueorangutan package

File: engine/schema/src/test/java/com/cloud/upgrade/DatabaseUpgradeCheckerTest.java [hunk 1]
```diff
@@ -165,9 +167,41 @@ public void testCalculateUpgradePathUnkownDbVersion() {
         final DatabaseUpgradeChecker checker = new DatabaseUpgradeChecker();
         final DbUpgrade[] upgrades = checker.calculateUpgradePath(dbVersion, currentVersion);
         assertNotNull(upgrades);
-        assertTrue(upgrades.length == 1);
-        assertTrue(upgrades[0] instanceof NoopDbUpgrade);
+        assertTrue(upgrades.length == 2);
```
Comment (DaanHoogland) [inline engine/schema/src/test/java/com/cloud/upgrade/DatabaseUpgradeCheckerTest.java:170]:
```suggestion
        assertEquals("We should have 2 upgrade steps", 2, upgrades.length);
```

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@rohityadavcloud @nvazquez @borisstoyanov can you review? I think this will be hard to test untill we reach a point in time where it is useful, but it has unit tests (added). Maybe give extra attention to the assertions in those during review!?

File: engine/schema/pom.xml [hunk 2]
```diff
@@ -57,6 +57,18 @@
             <artifactId>ini4j</artifactId>
             <version>${cs.ini.version}</version>
         </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter</artifactId>
+            <version>RELEASE</version>
```
Comment (rohityadavcloud) [inline engine/schema/pom.xml:63]: should we specify some version here? @DaanHoogland

File: engine/schema/src/main/java/com/cloud/upgrade/DatabaseVersionHierarchy.java [hunk 3]
```diff
@@ -131,7 +131,7 @@ private CloudStackVersion getRecentVersion(final CloudStackVersion fromVersion)
         return toList()
                  .reverse()
                  .stream()
-                 .filter(version -> fromVersion.compareTo(version) < 0)
```
Comment (rohityadavcloud) [inline engine/schema/src/main/java/com/cloud/upgrade/DatabaseVersionHierarchy.java:134]: Should we just remove the last digit (from a.b.c.d; remove .d) in the upgrade paths map and the logic that obtains current ACS version (to report a.b.c only to DB upgrade path processors?)

**Review by rohityadavcloud — APPROVED** [review]
Needs testing but LGTM if there are no regression tests

File: engine/schema/src/main/java/com/cloud/upgrade/DatabaseVersionHierarchy.java (hunk 3 above)
Comment (DaanHoogland) [reply engine/schema/src/main/java/com/cloud/upgrade/DatabaseVersionHierarchy.java:134]: unfortunately a lot of four digit versions are already in the DBs around the world. In addition to that, we might want to add a DB patch in a security release at some pioint in which case we want to take the fourth digit into account.

**Comment by DaanHoogland** [comment]
@blueorangutan package

File: engine/schema/src/test/java/com/cloud/upgrade/DatabaseVersionHierarchyTest.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+import com.cloud.upgrade.dao.Upgrade41710to41800;
+import com.cloud.upgrade.dao.Upgrade481to490;
+import org.apache.cloudstack.utils.CloudStackVersion;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import java.io.InputStream;
+import java.sql.Connection;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Comment (weizhouapache) [inline engine/schema/src/test/java/com/cloud/upgrade/DatabaseVersionHierarchyTest.java:35]: @DaanHoogland 
what's the difference between this and `Assert.assertEquals` ?

Comment (DaanHoogland) [reply engine/schema/src/test/java/com/cloud/upgrade/DatabaseVersionHierarchyTest.java:35]: this is junit5 as opposed to junit4

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by weizhouapache — APPROVED** [review]

**Comment by weizhouapache** [comment]
@blueorangutan package

File: engine/schema/src/test/java/com/cloud/upgrade/DatabaseVersionHierarchyTest.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+package com.cloud.upgrade;
+
+import com.cloud.upgrade.dao.DbUpgrade;
+import com.cloud.upgrade.dao.Upgrade41110to41120;
+import com.cloud.upgrade.dao.Upgrade41120to41130;
+import com.cloud.upgrade.dao.Upgrade41120to41200;
+import com.cloud.upgrade.dao.Upgrade41500to41510;
+import com.cloud.upgrade.dao.Upgrade41510to41520;
+import com.cloud.upgrade.dao.Upgrade41520to41600;
+import com.cloud.upgrade.dao.Upgrade41710to41800;
```
Comment (DaanHoogland) [inline engine/schema/src/test/java/com/cloud/upgrade/DatabaseVersionHierarchyTest.java:26]:
```suggestion
import com.cloud.upgrade.dao.Upgrade41720to41800;
```

File: engine/schema/src/test/java/com/cloud/upgrade/DatabaseVersionHierarchyTest.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+                .next("4.8.2.0" , new Upgrade481to490())
+                .next("4.9.10.11"   , new DummyUpgrade())
+                .next("4.11.1.0", new Upgrade41110to41120())
+                .next("4.11.2.0", new Upgrade41120to41130())
+                .next("4.11.3.0", new Upgrade41120to41200())
+                .next("4.15.0.0", new Upgrade41500to41510())
+                .next("4.15.1.0", new Upgrade41510to41520())
+                .next("4.15.2.0", new Upgrade41520to41600())
+                .next("4.15.4", new DummyUpgrade())
+                .next("4.17.1.0", new Upgrade41710to41800());
```
Comment (DaanHoogland) [inline engine/schema/src/test/java/com/cloud/upgrade/DatabaseVersionHierarchyTest.java:93]:
```suggestion
                .next("4.17.2.0", new Upgrade41710to41800());
```

**Review by DaanHoogland — COMMENTED** [review]
semantic conflict :(

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Pull request merged**